pkg/api/vmtemplate/store.go

```diff
@@ -0,0 +1,119 @@
+package vmtemplate
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/vm/pkg/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/rancher/apiserver/pkg/types"
+	ctlvmv1alpha1 "github.com/rancher/vm/pkg/generated/controllers/vm.cattle.io/v1alpha1"
+	"github.com/rancher/vm/pkg/ref"
+)
+
+const (
+	vmTemplateLabel = "vmtemplate.vm.cattle.io/vmTemplateID"
+)
+
+type wrapStore struct {
+	types.Store
+	vmTemplateCache        ctlvmv1alpha1.VMTemplateCache
+	vmTemplateVersions     ctlvmv1alpha1.VMTemplateVersionClient
+	vmTemplateVersionCache ctlvmv1alpha1.VMTemplateVersionCache
+	mu                     sync.Mutex
+}
+
+func (s *wrapStore) Create(request *types.APIRequest, schema *types.APISchema, data types.APIObject) (types.APIObject, error) {
+	ns := data.Data().String("metadata", "namespace")
+	if ns == "" {
+		data.Data().SetNested(config.Namespace, "metadata", "namespace")
+	}
+
+	vmTemplateID := data.Data().String("spec", "vmTemplateId")
+	if vmTemplateID == "" {
+		return types.APIObject{}, errors.New("VMTemplateId is empty")
+	}
+
+	if err := s.setOwnerRefAndLabel(request, data); err != nil {
+		return types.APIObject{}, err
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	_, templateName := ref.Parse(vmTemplateID)
+	sets := labels.Set{
+		vmTemplateLabel: templateName,
+	}
+	existedVersions, err := s.vmTemplateVersions.List(ns, metav1.ListOptions{LabelSelector: sets.AsSelector().String()})
+	if err != nil {
+		return types.APIObject{}, err
+	}
+
+	currentVersion := len(existedVersions.Items) + 1
+	data.Data().SetNested(currentVersion, "status", "versionNumber")
+	data.Data().SetNested(templateName+"-", "metadata", "generateName")
+
+	return s.Store.Create(request, request.Schema, data)
+}
+
+func (s *wrapStore) Update(request *types.APIRequest, schema *types.APISchema, data types.APIObject, id string) (types.APIObject, error) {
+	return types.APIObject{}, errors.New("Update vmTemplateVersion is not supported")
+}
+
+func (s *wrapStore) Delete(request *types.APIRequest, schema *types.APISchema, id string) (types.APIObject, error) {
+	if err := s.canDeleteTemplateVersion(id); err != nil {
+		return types.APIObject{}, err
+	}
+
+	return s.Store.Delete(request, request.Schema, id)
+}
+
+func (s *wrapStore) canDeleteTemplateVersion(id string) error {
+	ns, name := ref.Parse(id)
+	vr, err := s.vmTemplateVersionCache.Get(ns, name)
+	if err != nil {
+		return err
+	}
+
+	vtNS, vtname := ref.Parse(vr.Spec.VMTemplateID)
+	vt, err := s.vmTemplateCache.Get(vtNS, vtname)
+	if err != nil {
+		return err
+	}
+
+	if vt.Spec.DefaultVersionID == id {
+		return errors.New("Cannot delete default templateVersion")
```

Comment:
return an `apierror.APIError` with a proper response code.
`default`->`the default`

Reply:
updated